src/node/byzantine_identity.h

```diff
@@ -0,0 +1,331 @@
+    std::unordered_map<kv::ReconfigurationId, ByzantineIdentity> byids;
+
+  public:
+    ByzantineIdentitiesHook(
+      kv::Version version_, const ByzantineIdentities::Write& w) :
+      version(version_)
+    {
+      for (const auto& [rid, opt_bid] : w)
+      {
+        if (opt_bid.has_value())
+        {
+          LOG_DEBUG_FMT(
+            "New Byzantine network identity, valid for configuration #{}.",
+            rid);
+          byids.try_emplace(rid, opt_bid.value());
+        }
+      }
+    }
+
+    void call(kv::ConfigurableConsensus* consensus) override
+    {
+      for (auto& [rid, id] : byids)
+      {
+        consensus->add_identity(version, rid, id);
+      }
+    }
+  };
+
+  class ByzantineIdentityTracker
+  {
+  public:
+    enum class SessionState
+    {
+      STARTED,
+      FINISHED
+    };
+
+    class Session
+    {
+    public:
+      SessionState state = SessionState::STARTED;
+      kv::NetworkConfiguration config;
+
+      Session(const kv::NetworkConfiguration& config_) : config(config_) {}
+    };
+
+    ByzantineIdentityTracker(
+      std::shared_ptr<aft::State> shared_state_,
+      std::shared_ptr<enclave::RPCMap> rpc_map_,
+      crypto::KeyPairPtr node_sign_kp_,
+      const crypto::Pem& node_cert_) :
+      shared_state(shared_state_),
+      rpc_map(rpc_map_),
+      node_sign_kp(node_sign_kp_),
+      node_cert(node_cert_)
+    {}
+
+    virtual ~ByzantineIdentityTracker() {}
+
+    void add_network_configuration(const kv::NetworkConfiguration& config)
+    {
+      network_configs[config.rid] = config;
+    }
+
+    void reshare(const kv::NetworkConfiguration& config)
+    {
+      auto rid = config.rid;
+      LOG_DEBUG_FMT("Identities: start resharing for configuration #{}", rid);
+      sessions.emplace(rid, Session(config));
+
+      auto msg = std::make_unique<threading::Tmsg<UpdateIdentityTaskMsg>>(
+        update_identity_cb, rid, rpc_map, node_sign_kp, node_cert);
+
+      threading::ThreadMessaging::thread_messaging.add_task(
+        threading::ThreadMessaging::get_execution_thread(
+          threading::MAIN_THREAD_ID),
+        std::move(msg));
+    }
+
+    kv::ReconfigurationId find_reconfiguration(
+      const kv::Configuration::Nodes& nodes) const
+    {
+      // We're searching for a configuration with the same set of nodes as
+      // `nodes`. We currently don't have an easy way to look up the
+      // reconfiguration ID, which would make this unnecessary.
+      for (auto& [rid, nc] : network_configs)
+      {
+        bool have_all = true;
+        for (auto& [nid, byid] : nodes)
+        {
+          if (nc.nodes.find(nid) == nc.nodes.end())
+          {
+            have_all = false;
+            break;
+          }
+        }
+        if (have_all)
+        {
+          return rid;
+        }
+      }
+      return -1;
+    }
+
+    bool have_identity_for(
+      kv::ReconfigurationId rid, ccf::SeqNo commit_idx) const
+    {
+      if (rid == (kv::ReconfigurationId)-1)
```

Comment:
can (kv::ReconfigurationId)-1 be made into a const or something similar?

Reply:
It is. That was changed in another PR that's been merged into master, you may be reviewing an old un-rebased version of this PR?

Reply:
Ah, no, I was wrong, I missed on occurrence. Thanks!